### ISSUE TITLE
Issue with binding User Managed Spanner Service in PCF in spring cloud Hoxton

### ISSUE THREAD

**Issue by hari79** [spring-cloud/spring-cloud-gcp#2234]
we have a spring cloud app which is deployed on GCP thru PCF
our app has a service binding in pcf where GCP spanner instance is automatically bound by PCF in startup
- if my spanner service is added from PCF marketplace it works
- but if i create a user managed service out of the same the app does not start up as spanner binding fails

**Comment by meltsufin** [comment]
What's the difference with the user-managed Spanner service?
What does it look like in `VCAP_SERVICES`?

**Comment by hari79** [comment]
@meltsufin this is how the vcap_services looks like for the user managed service.a user managed service might point to a shared spanner service that is used across orgs or foundations

{
  "staging_env_json": {},
  "running_env_json": {},
  "system_env_json": {
    "VCAP_SERVICES": {
      "user-provided": [
        {
          "label": "user-provided",
          "name": "supply-spanner-service",
          "tags": [],
          "instance_name": "supply-spanner-service",
          "binding_name": null,
          "credentials": {
            "Email": "pcf-binding-abcde@pcf-stg-xx-xxxx.iam.gserviceaccount.com",
            "Name": "pcf-binding-abcde",
            "PrivateKeyData": "xxxxxxxxx",
            "UniqueId": "xxxxxxxxxx",
            "instance_id": "pcf-sb-1-1xxxxxxxxxx"
          },
          "syslog_drain_url": "",
          "volume_mounts": []
        }
      ]
    }
  },
  "application_env_json": {
    "VCAP_APPLICATION": {
      "cf_api": "https://api.cf01.useast4.xxxxxxx.cloud",
      "limits": {
        "fds": 16384
      },
      "application_name": "service-intelligence-api-bind-working",
      "application_uris": [
        "service-intelligence-api-bind-test.cf01.useast4.xxxxxxx.cloud"
      ],
      "name": "service-intelligence-api-bind-working",
      "space_name": "intelligence",
      "space_id": "xxxxxxxxxx",
      "uris": [
        "service-intelligence-api-bind-test.cf01.useast4.xxxxxxx.cloud"
      ],
      "users": null,
      "application_id": "xxxxxxxxxxxxxxxxx"
    }
  }
}

**Comment by hari79** [comment]
this is the error in startup

Cannot resolve reference to bean 'spannerTemplate' while setting bean property 'spannerTemplate'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.cloud.gcp.autoconfigure.spanner.GcpSpannerAutoConfiguration$CoreSpannerAutoConfiguration': Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.gcp.autoconfigure.spanner.GcpSpannerAutoConfiguration$CoreSpannerAutoConfiguration]: Constructor threw exception; nested exception is java.io.IOException: The Application Default Credentials are not available. They are available if running in Google Compute Engine. Otherwise, the environment variable GOOGLE_APPLICATION_CREDENTIALS must be defined pointing to a file defining the credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.

**Comment by meltsufin** [comment]
I'm not very familiar with the user-provided option in PCF, but looking at our `GcpCloudFoundryEnvironmentPostProcessor` there's no handling for it. In fact, the `VCAP_SERVICES` you pasted doesn't even identify your user-provided service as a Cloud Spanner instance. How can we make the connection that it is a configuration for Spanner?

@artembilan Can you connect us to someone who is more familiar with PCF to chime in on this question?

**Comment by scottfrederick** [comment]
It looks like detection of GCP services is only done by [checking the `label` field](https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-cloudfoundry/src/main/java/org/springframework/cloud/gcp/cloudfoundry/GcpCloudFoundryEnvironmentPostProcessor.java#L127) from `VCAP_SERVICES` . In the case of a user-provided service instance, the `label` is always `user-provided`. 

> How can we make the connection that it is a configuration for Spanner?

One suggestion is to also inspect the list of `tags` provided for each service in `VCAP_SERVICES`. For example:

```
List<CfService> serviceBindings = cfEnv.findServicesByLabel(cfServiceName);

if (serviceBindings == null) {
    serviceBindings = cfEnv.findServicesByTag(cfServiceName);
}

if (serviceBindings == null) {
    return properties;
}
```

Then you could instruct users to include the appropriate tag when creating the user-provided service instance, for example: 

```
cf create-user-provided-service supply-spanner-service -t "google-spanner" -p {credentials...}
```

**Comment by meltsufin** [comment]
@scottfrederick Thanks a lot for the suggestion! We can definitely implement this support with the tags.

**Comment by dzou** [comment]
Hey @hari79 @scottfrederick - I made a PR to process the tags as you suggested: https://github.com/spring-cloud/spring-cloud-gcp/pull/2238

It will now check the PCF name for the google service as a tag for the app and so deploying it with the tag should configure the library correctly now. Let me know if this is a satisfactory solution for you.

**Comment by hari79** [comment]
@dzou thank you for promptly adding the feature. appreciate it

**Comment by hari79** [comment]
@dzou should the tags look like this for the binding to work?

          "name": "supply-spanner-service",
          "tags": ["supply-spanner-service"],

**Comment by dzou** [comment]
@hari79 - No problem.

1. Not quite, you'll need the tags like this:

```
"tags": ["google-spanner"]
```

2. Since the change is new, you'll have to rely on the snapshot version of our library to begin using the change.

Might take a day for the change to be propagated to the snapshot; it will be available in version `1.2.3.BUILD-SNAPSHOT`.

Also will have to add the following repo to your maven build file:

```
<repository>
	<id>spring-snapshots</id>
	<name>Spring Snapshots</name>
	<url>https://repo.spring.io/libs-snapshot-local</url>
	<snapshots>
		<enabled>true</enabled>
	</snapshots>
</repository>
```
